java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -35,41 +33,80 @@ import kotlinx.coroutines.withContext
  */
 class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     val storageKey: StorageKey,
-    val backingType: Type,
-    val callbackFactory: (String) -> ProxyCallback<Data, Op, T>
+    val backingType: Type
 ) {
     private val storeMutex = Mutex()
     private val log = TaggedLog { "DirectStoreMuxer" }
+    private val proxyManager = RandomProxyCallbackManager<Data, Op, T>(
+        "backing",
+        Random
+    )
 
     // TODO(b/158262634): Make this CacheMap Weak.
     /* internal */ val stores = LruCacheMap<String, StoreRecord<Data, Op, T>>(
         50,
         livenessPredicate = { _, sr -> !sr.store.closed }
-    ) { _, sr -> closeStore(sr) }
+    ) { muxId, sr -> closeStore(muxId, sr) }
+
+    // Keeps track of all the stores a callbackId is registered to. Used when removing an
+    // observer (off method), as that requires unregistering to each store it is registered with.
+    private val callbackIdToMuxIdMap = mutableMapOf<Int, MutableSet<String>>()
+
+    fun on(callback: ProxyCallback<Data, Op, T>, callbackToken: Int? = null): Int {
+        // Direct Store Muxer must generate the callback token to ensure no duplicates.
+        check(callbackToken == null) { "token must be generated by Direct Store Muxer" }
+        synchronized(proxyManager) {
+            val callbackId = proxyManager.register(callback, callbackToken)
+            callbackIdToMuxIdMap[callbackId] = mutableSetOf()
+            return callbackId
+        }
+    }
+
+    fun off(callbackToken: Int) {
+        synchronized(proxyManager) {
+            for (muxId in callbackIdToMuxIdMap.getOrElse(callbackToken, { mutableSetOf() })) {
```

Comment:
Tiny nit... lambdas that are the last param in a function call go outside the closing paren.

```suggestion
            for (muxId in callbackIdToMuxIdMap.getOrElse(callbackToken) { mutableSetOf() }) {
```